File: frontend/src/lib/api/deletion.ts
/**
 * API client functions for account deletion.
 */
import { api } from './client';

// =============================================================================
// Types
// =============================================================================

export interface DeletionRequestCreate {
  reason?: string;
  export_data?: boolean;
}

export interface DeletionInitiatedResponse {
  message: string;
  deletion_request_id: string;
  scheduled_deletion_at: string;
  confirmation_email_sent: boolean;
  grace_period_days: number;
}

export interface DeletionConfirmRequest {
  password: string;
  confirmation_token: string;
}

export interface DeletionConfirmedResponse {
  message: string;
  deletion_request_id: string;
  scheduled_deletion_at: string;
  status: string;
}

export interface DeletionCancelledResponse {
  message: string;
  deletion_request_id: string;
  cancelled_at: string;
}

export interface DeletionRequest {
  id: string;
  user_id: string;
  status: 'pending' | 'confirmed' | 'executed' | 'cancelled';
  requested_at: string;
  scheduled_deletion_at: string;
  confirmed_at?: string;
  cancelled_at?: string;
  data_export_requested: boolean;
  days_remaining: number;
  can_cancel: boolean;
}

export interface DeletionStatusResponse {
  has_pending_deletion: boolean;
  deletion_request?: DeletionRequest;
}

// =============================================================================
// API Functions
// =============================================================================

/**
 * Request account deletion with 7-day grace period.
 */
export async function requestDeletion(
  data: DeletionRequestCreate
): Promise<DeletionInitiatedResponse> {
  const response = await api.post<DeletionInitiatedResponse>(
    '/users/me/request-deletion',
    data
  );
  return response;
}

/**
 * Confirm account deletion with password and token.
 */
export async function confirmDeletion(
  data: DeletionConfirmRequest
): Promise<DeletionConfirmedResponse> {
  const response = await api.post<DeletionConfirmedResponse>(
    '/users/me/confirm-deletion',
    data
  );
  return response;
}

/**
 * Cancel a pending or confirmed deletion request.
 */
export async function cancelDeletion(): Promise<DeletionCancelledResponse> {
  const response = await api.delete<DeletionCancelledResponse>(
    '/users/me/cancel-deletion'
  );
  return response;
}

/**
 * Get the current deletion status.
 */
export async function getDeletionStatus(): Promise<DeletionStatusResponse> {
  const response = await api.get<DeletionStatusResponse>(
    '/users/me/deletion-status'
  );
  return response;
}

File: backend/app/schemas/deletion.py
"""Pydantic schemas for account deletion flow."""
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class DeletionRequestCreate(BaseModel):
    """Request to initiate account deletion."""

    reason: str | None = Field(
        default=None,
        max_length=1000,
        description="Optional reason for deletion",
    )
    export_data: bool = Field(
        default=False,
        description="Request data export before deletion",
    )


class DeletionConfirmRequest(BaseModel):
    """Request to confirm account deletion with password."""

    password: str = Field(
        ...,
        min_length=1,
        description="Current password for verification",
    )
    confirmation_token: UUID = Field(
        ...,
        description="Token received in confirmation email",
    )


class DeletionCancelRequest(BaseModel):
    """Request to cancel a pending deletion."""

    pass  # No fields needed, authenticated user is implicit


class DeletionRequestResponse(BaseModel):
    """Response containing deletion request details."""

    id: UUID
    user_id: UUID | None  # NULL after user deletion (audit records)
    status: str
    requested_at: datetime
    scheduled_deletion_at: datetime
    confirmed_at: datetime | None = None
    cancelled_at: datetime | None = None
    data_export_requested: bool
    days_remaining: int = Field(
        ...,
        description="Days until permanent deletion (0 if already executed)",
    )
    can_cancel: bool = Field(
        ...,
        description="Whether the request can still be cancelled",
    )

    model_config = {"from_attributes": True}


class DeletionInitiatedResponse(BaseModel):
    """Response when deletion request is initiated."""

    message: str
    deletion_request_id: UUID
    scheduled_deletion_at: datetime
    confirmation_email_sent: bool
    grace_period_days: int


class DeletionConfirmedResponse(BaseModel):
    """Response when deletion is confirmed."""

    message: str
    deletion_request_id: UUID
    scheduled_deletion_at: datetime
    status: str


class DeletionCancelledResponse(BaseModel):
    """Response when deletion is cancelled."""

    message: str
    deletion_request_id: UUID
    cancelled_at: datetime


class DeletionStatusResponse(BaseModel):
    """Response containing current deletion status."""

    has_pending_deletion: bool
    deletion_request: DeletionRequestResponse | None = None


class DeletionSummary(BaseModel):
    """Summary of what was deleted."""

    user_deleted: bool
    students_deleted: int
    notes_deleted: int
    flashcards_deleted: int
    sessions_deleted: int
    ai_interactions_deleted: int
    files_deleted: int
    total_records_deleted: int
